### ISSUE TITLE
Support creating tables from SELECT statements

### ISSUE THREAD

**Issue by JLDLaughlin** [MaterializeInc/materialize#5266]
Right now, we only support creating tables statically like:
```
CREATE TABLE example (a int, b text NOT NULL)
```

It would be great to support creating tables from `SELECT` statements, like:
```
CREATE TABLE example AS SELECT 1
```

Postgres supports this:
```
jessicalaughlin=# CREATE TABLE example AS SELECT 1;
SELECT 1
jessicalaughlin=# SELECT * FROM example;
 ?column?
----------
        1
(1 row)
```

This would make our table functionality more flexible. As a bonus, it would allow us to easily create tables using [dbt](https://www.getdbt.com/) (dbt expects [materializations](https://docs.getdbt.com/docs/building-a-dbt-project/building-models/materializations/) to be created from `SELECT` statements).

**Comment by benesch** [comment]
💯 agree that this would be super nice! Just want to flag that it is not as easy as it might seem. /cc @mjibson 

The tl;dr is that even with our fancy new transaction support, we're going to limit transactions to exactly one of: catalog changes, reads, and writes. `CREATE TABLE ... AS SELECT ...` is a single statement that manages to do all three of those things.

There _are_ ways to make it work with some aggressive special casing, but we should proceed with caution.

**Comment by rjnn** [comment]
I wonder what use-case would require a `CREATE TABLE x AS SELECT y FROM z`? Wouldn't `CREATE MATERIALIZED VIEW x AS SELECT y FROM z` suffice?

I suppose the one thing you might want is for `x` to _not_ change as `z` changes..?

**Comment by benesch** [comment]
> I suppose the one thing you might want is for `x` to _not_ change as `z` changes..?

Yep, that's exactly the use case I was imagining. 

But perhaps dbt would be satisfied enough with a materialized view here instead of a table! @JLDLaughlin, do you know if that's the case?

**Comment by JLDLaughlin** [comment]
> But perhaps dbt would be satisfied enough with a materialized view here instead of a table! @JLDLaughlin, do you know if that's the case?

Yes! That's actually what's implemented now. I think there's a UX mismatch of telling users they're creating a table and creating a materialized view instead - I don't expect every user will read the documentation carefully. This isn't super high priority, just something I wanted to suggest!

**Comment by benesch** [comment]
> > But perhaps dbt would be satisfied enough with a materialized view here instead of a table! @JLDLaughlin, do you know if that's the case?
> 
> Yes! That's actually what's implemented now. I think there's a UX mismatch of telling users they're creating a table and creating a materialized view instead - I don't expect every user to read the documentation carefully. This isn't super high priority, just something I wanted to suggest!

Ah, got it! Agreed that this is something we should do. It's just hard.

**Comment by rjnn** [comment]
What if we did something that's the equivalent of `CREATE MATERIALIZED VIEW x AS SELECT y FROM z AS OF t`?

I know `AS OF` is really complicated and there are 12 different meanings, but like that could "freeze" `x` to `t`?

**Comment by mjibson** [comment]
From a transactional standpoint, if we support this (which I'm for), we would need to make sure that there is no time at which the table (or view if it ends up being that) exists but is also empty or only partially correct. My guess is that's not too hard: we can do all the dataflow work first and then add the catalog entry last? I'm not familiar enough with the catalog to know if that's possible.

**Comment by benesch** [comment]
We'd also need this to support the "in-warehouse" mode of reverse eTL tools. This is Hightouch's "lightning" sync, Census's "advanced" sync, and the _only_ sync option for Rudderstack.

We're not too worried about supporting Hightouch's lightning sync or Census's advanced sync, as 1) we have a working basic connector with both of these tools, and 2) the long term goal is to build a stream-native advanced sync mode that uses [`CREATE SUBSCRIPTION`](https://github.com/MaterializeInc/materialize/issues/17541).

It's much more painful for Rudderstack though, as there is no basic version of the sync to fall back on, and so we don't have a Rudderstack rETL story at all right now.
